Serialize config with protobuf

IrinaShkviro opened the pull request:
Replaced cereal with protobuf for config file and removed unused fields from it.

New config file is stored in `config.textpb`, and stores only values that are different from default (false for bool, zero for numeric types, empty string for strings and empty for repeated fields).
I named a new target `OrbitFileProtos` initially, but would be glad to improve naming.

dimitry- commented on OrbitCore/Params.cpp:
```diff
@@ -4,58 +4,28 @@
-#include <algorithm>
 #include <fstream>
 
+#include <google/protobuf/io/zero_copy_stream_impl.h>
```
we use "" for absl, lets use the same for protobuf

IrinaShkviro replied:
Done, thanks!

dimitry- commented on OrbitCore/Params.cpp:
```diff
@@ -67,12 +37,12 @@ bool Params::Save() {
-    cereal::XMLOutputArchive archive(file);
-    archive(cereal::make_nvp("Params", *this));
+  google::protobuf::io::OstreamOutputStream output_stream(&file);
+  if (google::protobuf::TextFormat::Print(config, &output_stream)) {
```
nitpick:

if (! .. ) {
   ...
   return false;
}

return true;

dimitry- replied:
btw what is output format here?

IrinaShkviro replied:
The sample of `config.textpb` file:
```
track_context_switches: true
track_sampling_events: true
unreal_support: true
start_paused: true
find_file_and_line_info: true
font_size: 14

dimitry- commented on OrbitCore/Params.cpp:
```diff
@@ -87,12 +57,11 @@ bool Params::Load() {
-    cereal::XMLInputArchive archive(file);
-    archive(*this);
+  google::protobuf::io::IstreamInputStream input_stream(&file);
+  if (google::protobuf::TextFormat::Parse(&input_stream, &config)) {
```
same here.

IrinaShkviro replied:
Updated PR

dpallotti commented on protos/config.proto:
```diff
@@ -0,0 +1,22 @@
+
+syntax = "proto3";
+
+message Config {
```
As this will have to be compatible across versions, I would try to minimize the number of fields that we define. The vast majority of these are used in the code, but don't have a corresponding ui and that code effectively does nothing. So I would prefer to get rid of that code rather than adapting new protos to it.

IrinaShkviro replied:
Makes sense, thank you so much, will do!

IrinaShkviro commented:
Some analysis on config file: all fields could be divided into several categories

1. Get the field value in one or several places, and set only in UI in `--enable_stale_features` mode:

```
unreal_support (true);
bool allow_unsafe_hooking (false);
bool hook_output_debug_string (false);
```

2. Get the field value in one or several places, and there's no places that change the value:
```
bool start_paused (true);
bool find_file_and_line_info (true);
float font_size (14.f);
string arguments ("");
string working_directory ("");
string process_filter ("");
```

3. Get and set the field value only in UI in `--enable_stale_features` mode, and there is no code that depends on it:
```
bool track_context_switches;
bool track_sampling_events;
bool upload_dumps_to_server;
```

4. Only change the field value, don't get the field value from the code:
```
repeated string pdb_history;
```

I removed 1-3 categories for now, and unsure about 4th. Would be glad to hear opinions on that.

dpallotti commented:
> I removed 1-3 categories for now, and unsure about 4th. Would be glad to hear opinions on that.

As I've said, I'm for removing as much unused code as possible, otherwise we'll be having the same conversation again on why which fields are used where: http://go/deadcode
So about category 4, I'm for removing it&mdash;the name is already misleading as Pdb is Windows.
I would also be for removing category 2 and simply use a constant for now in the code that uses it (or just remove that code if it's easy to do so)&mdash;but I know you haven't asked about category 2. 😄

IrinaShkviro commented:
> > I removed 1-3 categories for now, and unsure about 4th. Would be glad to hear opinions on that.
> 
> As I've said, I'm for removing as much unused code as possible, otherwise we'll be having the same conversation again on why which fields are used where: http://go/deadcode
> So about category 4, I'm for removing it—the name is already misleading as Pdb is Windows.
> I would also be for removing category 2 and simply use a constant for now in the code that uses it (or just remove that code if it's easy to do so)—but I know you haven't asked about category 2. 😄

Actually, I wanted to ask about all the categories :) About the category 2: I removed it also, and code that uses it (except the `font_size` - I added the constant for it).
